### DISCUSSION TITLE
Error: dia.ElementView: markup required - for custom shapes after graph.fromJSON

### DISCUSSION THREAD

**Discussion by frnora** [clientIO/joint#2099] [Q&A] [answered]
Hi there,

I have read a few discussions about this but I am still not quite sure what's going on with my example.
I have some custom shapes and views - they all work nicely until I try to load them using `graph.fromJSON`. When I do that, I get `Error: dia.ElementView: markup required`. 

This is a minimalistic example of what I am doing. I have a custom element, something like this:

    import { dia } from "@clientio/rappid";

    class Base extends dia.Element {

        markup = [{
            tagName: 'path',
            selector: 'body'
        },
            // ...
        ]

        defaults() {
            return {
                ...super.defaults,
                type: 'myShapes.Base',
                size: { width: 120, height: 60 },
                attrs: {
                    body: {
                        fill: '#FFFFFF',
                        d: 'some_path'
                    }
                    // ...
                }
            }
        }

        // some other setters/getters
    }

    export default Base;

This element has an associated view:

    import { dia } from "@clientio/rappid";
    import './BaseView.css';

    const BaseView = dia.ElementView.extend({

        initialize() {
             dia.ElementView.prototype.initialize.apply(this, arguments);
             this.el.classList.add('my-el');
        }
        // other methods
    });

    export default BaseView;

I have a bunch of other elements extending this Base model and BaseView, nothing too complicated.

Now in my main app class, I import them all and then add them to the shapes object, paper and graph:

    import { dia, ui, shapes } from '@clientio/rappid';

    // ....

    Object.assign(shapes, {
          myShapes: {
                Base,
                BaseView,
                // rest...
          }
    });

    // ....

    this.#graph = new dia.Graph({ cellNamespace: shapes });

    // ....

    let paper = this.#paper = new dia.Paper({ model: this.#graph, cellViewNamespace: shapes });

    const base = new shapes.myShapes.Base({
        position: { x: 0, y: 0 },
    });
    this.#graph.addCell(base);

What am I missing? Why can't I load the JSON I previously saved with `graph.toJSON`? Thanks a lot for any feedback!

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by kumilingus** [comment]
Try to initialize the `markup` inside the `preinitialize()` lifecycle method.

```js
preinitialize() {
    this.markup = [{
        tagName: 'path',
        selector: 'body'
    },
        // ...
    ]
}
```

See the [example](https://jsfiddle.net/kumilingus/97dwxqv3/) and the [docs](https://backbonejs.org/#Model-preinitialize).

If this is not the problem then the problem lies somewhere in the `// rest...`.

**Comment by frnora** [comment]
Thanks for the reply, Roman.

That doesn't work either, still the same issue.

I removed everything I could, I am left with 2 files: App.js and TestEl.js. The TestEl.js is basically just copy-pasted Tooltip from the example you posted. So this is how it looks like:

App.js:
```js
import { dia, ui, shapes } from '@clientio/rappid';

import TestEl from './shapes/TestEl';

class App {

    #graph;
    #paper;
    #paperEl;
    #paperScroller;

    constructor(el) {
        Object.assign(shapes, {
            myShapes: {
                TestEl
            }
        });
        this.#initEls(el);
        this.#initGraph();
        this.#initPaper();
    }


    #initEls(el) {
        this.#paperEl = el.querySelector('.paper');
    }

    #initGraph() {
        this.#graph = new dia.Graph({ cellNamespace: shapes });
    }

    #initPaper() {
        let paper = this.#paper = new dia.Paper({
            model: this.#graph,
            cellViewNamespace: shapes,
            interactive: true,
            width: 10000,
            height: 10000,
            background: {
                color: '#fff'
            },
        });

        let paperScroller = this.#paperScroller = new ui.PaperScroller({
            autoResizePaper: false,
            scrollWhileDragging: true,
            cursor: 'grab',
            padding: 0,
            paper
        });
        this.#paperEl.appendChild(paperScroller.el);
        paperScroller.render();

        let t = new TestEl({
            position: { x: 0, y: 0 },
        });
        this.#graph.addCell(t);
    }

    get graph() {
        return this.#graph;
    }
}

export default App;
```

TestEl.js:
```js
import { dia } from "@clientio/rappid";

class TestEl extends dia.Element {

    defaults() {
        return {
            ...super.defaults,
            type: 'myShapes.TestEl',
            size: {
                width: 120,
                height: 40
            },
            attrs: {
                body: {
                    fill: '#ffffff',
                    stroke: '#333333',
                    strokeWidth: 2,
                    d: 'M 0 0 H calc(w) V calc(h) H 40 l -20 20 v -20 H 0 z',
                },
                label: {
                    text: 'Tooltip',
                    fontSize: 14,
                    fontFamily: 'sans-serif',
                    fill: '#333333',
                    textVerticalAnchor: 'middle',
                    textAnchor: 'middle',
                    x: 'calc(0.5*w)',
                    y: 'calc(0.5*h)',
                },
            }
        };
    }

    preinitialize() {
        this.markup = [{
            tagName: 'path',
            selector: 'body'
        }, {
            tagName: 'text',
            selector: 'label'
        }];
    }
}

export default TestEl;
```

And obviously the init code (in index.js):
```js
import App from './src/App';
import '@clientio/rappid/rappid.css';

let app = window.app = new App(document.getElementById('app'));
```

This is all just bundled up similarly how you doing it with your latest examples (I just basically copied the webpack config).

So this throws still the same error - missing markup. There must be something I am missing here...

**Comment by kumilingus** [comment, marked as answer]
I know 🤦 😄 
```js
  // the first parameter is graph model attributes, the other is the options.
  this.#graph = new dia.Graph({}, { cellNamespace: shapes });
  ```

**Reply by frnora** [reply]
Oh my... 😄 You are so right! Thanks a lot!
